charms/kserve-controller/requirements-unit.in

```
@@ -1,4 +1,22 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+# Please note this file introduces dependencies from the charm's requirements.in,
+# special attention must be taken when updating this or the other .in file to try
+# to avoid incompatibilities.
+# Rules for editing this file:
+#   * Removing a dependency that is no longer used in the unit test file(s)
+#     is allowed, and should not represent any risk.
+#   * Adding a dependency in this file means the dependency is directly used
+#     in the unit test files(s).
+#   * ALL python packages/libs used directly in the unit test file(s) must be
+#     listed here even if requirements.in is already adding them. This will
+#     add clarity to the dependency list.
+#   * Pinning a version of a python package/lib shared with requirements.in
+#     must not introduce any incompatibilities.
+charmed-kubeflow-chisme
 coverage
+lightkube
+ops
 pytest
 pytest-mock
--r requirements.txt
+-r requirements.in
```

This should be `requirements.txt`.  If we use `requirements.in` here, then our unit tests may get a different set of dependencies than our charm.  
```suggestion
-r requirements.txt
```

---

Ideally you don't do that, if you are in that scenario where you need a different version of a shared package between requirements.in and requirements-unit.in, you'll re-evaluate the reason you are pinning to such version. If the goal  is to have an environment closer to what the charm will have, then you should avoid at all costs introducing such changes.

---

The problem is the failure mechanism:
1. citing `requirements.txt` means that package incompatabilities will raise as failures during `pip-compile`.  They will be loud and clear, and can be resolved as soon as they come up
2. using `requirements.in` means that package incompatabilities **will not** raise as failures, instead quietly succeeding.  They will be especially hard to spot, and could lead to untested dependencies going into the charm

By doing (1) you know about the failure.  By doing (2), you don't.  imo that makes it clear `.txt` should be used

---

We can ignore this for now.  We will discuss later this week and, if we need a change we will push it separately.